packages/eslint-plugin/src/rules/no-void-expression.ts

```diff
@@ -0,0 +1,284 @@
+              let newReturnStmtText = `${returnValueText}; return;`;
+              if (returnStmt.parent?.type !== AST_NODE_TYPES.BlockStatement) {
+                // e.g. `if (cond) return console.error();`
+                // add braces if not inside a block
+                newReturnStmtText = `{ ${newReturnStmtText} }`;
+              }
+              return fixer.replaceText(returnStmt, newReturnStmtText);
+            },
+          });
+        }
+
+        // handle generic case
+        if (options.ignoreVoidOperator) {
+          // this would be reported by this rule btw. such irony
+          return context.report({
+            node,
+            messageId: 'invalidVoidExprWrapVoid',
+            suggest: [{ messageId: 'voidExprWrapVoid', fix: wrapVoidFix }],
+          });
+        }
+        context.report({
+          node,
+          messageId: 'invalidVoidExpr',
+        });
+      },
+    };
+
+    /**
+     * Inspects the void expression's ancestors and finds closest invalid one.
+     * By default anything other than an ExpressionStatement is invalid.
+     * Parent expressions which can be used for their short-circuiting behavior
+     * are ignored and their parents are checked instead.
+     * @param node The void expression node to check.
+     * @returns Invalid ancestor node if it was found. `null` otherwise.
+     */
+    function findInvalidAncestor(node: TSESTree.Node): TSESTree.Node | null {
+      const { parent } = node;
+
+      if (parent == null) {
+        // if there is no parent then let's say it's valid
+        return null;
+      }
+
+      if (parent.type === AST_NODE_TYPES.ExpressionStatement) {
+        // e.g. `{ console.log("foo"); }`
+        // this is always valid
+        return null;
+      }
+
+      if (parent.type === AST_NODE_TYPES.LogicalExpression) {
+        if (parent.right === node) {
+          // e.g. `x && console.log(x)`
+          // this is valid only if the next ancestor is valid
+          return findInvalidAncestor(parent);
+        }
+      }
+
+      if (parent.type === AST_NODE_TYPES.ConditionalExpression) {
+        if (parent.consequent === node || parent.alternate === node) {
+          // e.g. `cond ? console.log(true) : console.log(false)`
+          // this is valid only if the next ancestor is valid
+          return findInvalidAncestor(parent);
+        }
+      }
+
+      if (parent.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+        // e.g. `() => console.log("foo")`
+        // this is valid with an appropriate option
+        if (options.ignoreArrowShorthand) {
+          return null;
+        }
+      }
+
+      if (parent.type === AST_NODE_TYPES.UnaryExpression) {
+        if (parent.operator === 'void') {
+          // e.g. `void console.log("foo")`
+          // this is valid with an appropriate option
+          if (options.ignoreVoidOperator) {
+            return null;
+          }
+        }
+      }
```

Comment:
do you want to combine these?
```suggestion
      // e.g. `void console.log("foo")`
      // this is valid with an appropriate option
      if (
          parent.type === AST_NODE_TYPES.UnaryExpression &&
          parent.operator === "void" &&
          options.ignoreVoidOperator
      ) {
        return null;
      }
```

Reply:
I don't like when a single condition is broken on multiple lines like that so I sometimes make nested `if`s. Especially when I want to add a comment which only describes one code path but not the nested `if`. It also allows other developers to easily add other code paths when they find a case to handle (e.g. if typescript added a new unary operator which we also want to ignore but with separate option). I will leave it like that for now.

Reply:
all good, it's why I phrased it as "do you want to" and not "you should" or "you can".